File: leetcode/p00707/p00707.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None

class MyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def get(self, index: int) -> int:
        if self.size == 0 or index < 0 or index >= self.size:
            return -1

        current = self.head
        for i in range(index):
            current = current.next
        
        return current.val

    def addAtHead(self, val: int) -> None:
        node = Node(val)
        if self.size == 0:
            self.head = node
            self.tail = node
            self.size += 1
            return
        
        self.head.prev = node
        node.next = self.head
        self.head = node
        self.size += 1

    def addAtTail(self, val: int) -> None:
        node = Node(val)
        if self.size == 0:
            self.head = node
            self.tail = node
            self.size += 1
            return
        
        self.tail.next = node
        node.prev = self.tail
        self.tail = node
        self.size += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.size:
            return
        if index < 0:
            index = 0
        
        if index == 0:
            self.addAtHead(val)
            return
        
        if index == self.size:
            self.addAtTail(val)
            return
        
        node = Node(val)
        current = self.head
        for _ in range(0, index-1, 1):
            current = current.next
        
        node.next = current.next
        node.prev = current
        current.next.prev = node
        current.next = node
        self.size += 1

    def deleteAtIndex(self, index: int) -> None:
        if self.size == 0 or index < 0 or index >= self.size:
            return

        current = self.head
        for _ in range(index):
            current = current.next
        
        if current.prev != None:
            current.prev.next = current.next
        else:
            self.head = current.next
        
        if current.next != None:
            current.next.prev = current.prev
        else:
            self.tail = current.prev
        
        current.prev = None
        current.next = None
        
        self.size -= 1

if __name__ == '__main__':
    s = MyLinkedList()
    s.addAtHead(1)
    s.addAtTail(3)
    s.addAtIndex(1,2)
    assert s.get(1) == 2
    s.deleteAtIndex(1)
    assert s.get(1) == 3

    s = MyLinkedList()
    s.addAtHead(1)
    s.addAtTail(3)
    s.addAtIndex(1,2)
    assert s.get(1) == 2
    s.deleteAtIndex(0)
    assert s.get(0) == 2

File: leetcode/p00707/p00707.go
package p00707

type Node struct {
	Val  int
	Next *Node
	Prev *Node
}

type MyLinkedList struct {
	head *Node
	tail *Node
	size int
}

func Constructor() MyLinkedList {
	return MyLinkedList{
		head: nil,
		tail: nil,
		size: 0,
	}
}

func (this *MyLinkedList) Get(index int) int {
	if this.size == 0 || index < 0 || index >= this.size {
		return -1
	}

	current := this.head
	for i := 0; i < index; i++ {
		current = current.Next
	}

	return current.Val
}

func (this *MyLinkedList) AddAtHead(val int) {
	node := &Node{Val: val, Next: nil, Prev: nil}
	if this.size == 0 {
		this.head = node
		this.tail = node
		this.size++
		return
	}

	node.Next = this.head
	this.head.Prev = node
	this.head = node
	this.size++
}

func (this *MyLinkedList) AddAtTail(val int) {
	node := &Node{Val: val, Next: nil, Prev: nil}
	if this.size == 0 {
		this.head = node
		this.tail = node
		this.size++
		return
	}

	node.Prev = this.tail
	this.tail.Next = node
	this.tail = node
	this.size++
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index > this.size {
		return
	}
	if index < 0 {
		index = 0
	}
	if index == 0 {
		this.AddAtHead(val)
		return
	}
	if index == this.size {
		this.AddAtTail(val)
		return
	}

	current := this.head
	for i := 0; i < index-1; i++ {
		current = current.Next
	}

	node := &Node{Val: val, Next: nil, Prev: nil}
	node.Next = current.Next
	node.Prev = current
	current.Next.Prev = node
	current.Next = node
	this.size++
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if this.size == 0 || index < 0 || index >= this.size {
		return
	}

	current := this.head
	for i := 0; i < index; i++ {
		current = current.Next
	}

	if current.Prev != nil {
		current.Prev.Next = current.Next
	} else {
		this.head = current.Next
	}

	if current.Next != nil {
		current.Next.Prev = current.Prev
	} else {
		this.tail = current.Prev
	}

	current.Next = nil
	current.Prev = nil
	this.size--
}
